### PR TITLE
wip ITS-vertexer: Make cut on XY a configurable param

### PR DESCRIPTION
**Pull request by mconcas**

### CODE REVIEW

**Comment by mconcas** [comment]
@shahor02 : previously it was a cut on the XY position of the vertices to stay into the beamline.
Now it's a cut on the distance from beam to be expressed as `n*nsigma`.
I left the default configuration to mostly behave like earlier, modulo the beam shift (i.e.  n*nsigma=r_beampipe).

I'll then try to extract some meaningful value for sigma and then we should keep that and change only the other one (nsigma).
If it turns out to be too cumbersome we can drop one of them.

To be decided what will be the defaults for next data taking.

**Comment by mconcas** [comment]
Hi @shahor02, an update.
Namely the XY resolution computed on MC at the time of my PhD thesis was `sigmaX,Y=25`um and `sigmaZ=45`um.
Now, probably this is quite optimistic to be usable on real data. I'd opt for something more conservative both in sync and async.
Currently I think @davidrohr is testing with 300um cut.
Truth tobe told, already running with a 1mm cut on XY seems to tackle the tested issue, see:
Screenshot 2023-03-29 at 16 38 14
zoomed:
Screenshot 2023-03-29 at 16 38 42

Then, for some more refined studies I'd opt to run them on anchored MCs, if you agree.

File: Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx [hunk 1]
```diff
@@ -578,18 +577,17 @@ void VertexerTraits::computeVerticesInRof(int rofId,
   bool atLeastOneFound{false};
   for (int iCluster{0}; iCluster < nClusters; ++iCluster) {
     bool lowMultCandidate{false};
+    double beamDistance{sqrt((beamPosXY[0] - clusterLines[iCluster].getVertex()[0]) * (beamPosXY[0] - clusterLines[iCluster].getVertex()[0]) +
```
Comment (shahor02) [inline Detectors/ITSMFT/ITS/tracking/src/VertexerTraits.cxx:580]: would be faster to compare squared radii

**Review by shahor02 — COMMENTED** [review]
25 \mum tracklet resolution is indeed too optimistic, even for ideal MC. I would use ~500 \mum for vertex sigma (for the distance wrt the mean vertex from the CCDB) and ~100 \mum (configurable) for the tracklet sigma. And cut on the 
```
auto nsigmaCut = vertRadiusNsigmaCut*vertRadiusNsigmaCut*(vtxSigma*vtxSigma + trackletSigma*trackletSigma); // precalculated

auto dx = mTimeFrame->getBeamX() - mTimeFrame->getTrackletClusters(rofId)[iCluster].getVertex()[0];
auto dy = mTimeFrame->getBeamY() - mTimeFrame->getTrackletClusters(rofId)[iCluster].getVertex()[1];
double beamDistance2 = dx*dx + dy*dy;
if (beamDistance2 < nsigmaCut) {
 // accept
}
```
On your 2nd plot there still some outliers, at which radius they are?

**Comment by mconcas** [comment]
> 25 \mum tracklet resolution is indeed too optimistic, even for ideal MC. I would use ~500 \mum for vertex sigma (for the distance wrt the mean vertex from the CCDB) 

The 25um was essentially the RMS of the XY dist of vertices, but I agree sim conditions were probably less complex at that time. What I am changing in this PR is just a cut on the distance fo the verex to the beamline.

> and ~100 \mum (configurable) for the tracklet sigma. And cut on the
> 

I am not sure what you mean by tracklet resolution, currently the minimum disntance we allow for tracklets to be matched in a cluster is `mVertParams.paircut=400\um` and this very cut is the one that we apply for later matching of a third or further tracklet to the centroid of the cluster. Would you reduce this cut? I can surely have a try.

> ```
> auto nsigmaCut = vertRadiusNsigmaCut*vertRadiusNsigmaCut*(vtxSigma*vtxSigma + trackletSigma*trackletSigma); // precalculated
> 
> auto dx = mTimeFrame->getBeamX() - mTimeFrame->getTrackletClusters(rofId)[iCluster].getVertex()[0];
> auto dy = mTimeFrame->getBeamY() - mTimeFrame->getTrackletClusters(rofId)[iCluster].getVertex()[1];
> double beamDistance2 = dx*dx + dy*dy;
> if (beamDistance2 < nsigmaCut) {
>  // accept
> }
> ```
> 

> On your 2nd plot there still some outliers, at which radius they are?

I'll correlate them and let you know

**Comment by shahor02** [comment]
@mconcas sorry, I should have written "tracklet vertex" resolution...

**Comment by davidrohr** [comment]
@mconcas @shahor02 : I have retried 4GPU async reco with this PR and additional `--configKeyValues="ITSVertexerParam.vertRadiusSigma=0.005;ITSVertexerParam.vertRadiusNsigmaCut=6"`.
As expected, the large outliers in ITS tracking, ITS-TPC matching, and primarily vertexing time are gone.
So we should improve / tune these cuts, and then make it the default!

**Comment by mconcas** [comment]
> @mconcas @shahor02 : I have retried 4GPU async reco with this PR and additional `--configKeyValues="ITSVertexerParam.vertRadiusSigma=0.005;ITSVertexerParam.vertRadiusNsigmaCut=6"`. As expected, the large outliers in ITS tracking, ITS-TPC matching, and primarily vertexing time are gone. So we should improve / tune these cuts, and then make it the default!

Sure, since we are in async I'd do one run on MC, since efficiency control should be a concern there.

**Comment by shahor02** [comment]
@mconcas is not this ready to be merged?

**Comment by mconcas** [comment]
> @mconcas is not this ready to be merged?

not yet, I'm working on that. I'll finish it by today

**Comment by mconcas** [comment]
Closing in favour of https://github.com/AliceO2Group/AliceO2/pull/11142

**Pull request closed without merging**